### PR TITLE
Add priority downlink policy

### PR DESCRIPTION
**Pull request by ziyiz-amzn**
### Description of changes:
Adding mobile sdk/demo parity for the priority downlink policy. Builders will be able to assign priority to specific remote video streams.

### Testing done:
Tested simultaneously with 2 JS SDK clients and 1 iOS client. Turned video on/off several times for the JS SDK clients to test the new iOS downlink policy.

#### Manual test cases (add more as needed):

- [ ] Join meeting without CallKit
- [ ] Join meeting with CallKit as Incoming
- [ ] Join meeting with CallKit as Outgoing
- [ ] Leave meeting
- [ ] Rejoin meeting
- [ ] See metrics
- [ ] See attendee presence
- [ ] Send audio
- [ ] Receive audio
- [ ] Mute self
- [ ] Unmute self
- [ ] See local mute indicator when muted
- [ ] See remote mute indicator when other is muted
- [ ] Enable and disable local video
- [ ] Enable and disable remote video from remote side
- [ ] Send local video
- [ ] Pause and resume remote video
- [ ] Switch local camera
- [ ] Receive remote screen share

#### Integration validation (required before release)

- [ ] Unit tests pass
- [ ] Build SDK against simulator with release options
- [ ] Build SDK against device with release options
- [ ] Build SDK against simulator with debug options
- [ ] Build SDK against device with debug options
- [ ] Test Demo against simulator with SDK (debug build) in workspace
- [ ] Test Demo against device with SDK (debug build) in workspace
- [ ] Test DemoObjC against simulator with SDK (debug build) in workspace
- [ ] Test DemoObjC against device with SDK (debug build) in workspace
- [ ] Test Demo against simulator with SDK.framework (release build)
- [ ] Test Demo against device with SDK.framework (release build)
- [ ] Test DemoObjC against simulator with SDK.framework (release build)
- [ ] Test DemoObjC against device with SDK.framework (release build)

### Screenshots, if available:

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: AmazonChimeSDK/AmazonChimeSDK/audiovideo/AudioVideoControllerFacade.swift [hunk 1]
```diff
@@ -87,4 +87,26 @@ import Foundation
...
+    /// tile controller to indicate to the application that the tile should be unbound. To restart the flow of media, the source should be re-added by
+    /// including in `addedOrUpdated`. Note that videos no longer available in a meeting (i.e. listed in
+    /// `remoteVideoSourcesDidBecomeUnavailable` do not need to be removed, as they will be automatically unsubscribed from.
+    ///
+    /// Note that before this function is called for the first time, the client will automatically subscribe to all video sources.
+    /// However this behavior will cease upon first call (e.g. if there are 10 videos in the meeting, the controller will subscribe to all 10, however if
+    /// `updateVideoSourceSubscriptions` is called with a single video in `addedOrUpdated`, the client will unsubscribe from the other 9.
+    /// This automatic subscription behavior may be removed in future major version updates, builders should avoid relying on the logic
+    /// and instead explicitly call `updateVideoSourceSubscriptions` with the sources they want to receive.
```
Comment (hensmi-amazon) [inline AmazonChimeSDK/AmazonChimeSDK/audiovideo/AudioVideoControllerFacade.swift:107]: These are great

**Review by hensmi-amazon — APPROVED** [review]

File: AmazonChimeSDK/AmazonChimeSDK/internal/video/DefaultVideoClientController.swift [hunk 2]
```diff
@@ -229,6 +229,34 @@ extension DefaultVideoClientController: VideoClientDelegate {
             }
         }
     }
+    
+    public func remoteVideoSourcesDidBecomeAvailable(_ sourcesInternal: [RemoteVideoSourceInternal]) {
+        var sources = [RemoteVideoSource]()
```
Comment (hokyungh) [inline AmazonChimeSDK/AmazonChimeSDK/internal/video/DefaultVideoClientController.swift:234]: should you check for empty? We don't want to give callback when `sourcesInternal` is empty.

File: AmazonChimeSDK/AmazonChimeSDK/internal/video/DefaultVideoClientController.swift [hunk 3]
```diff
@@ -229,6 +229,34 @@ extension DefaultVideoClientController: VideoClientDelegate {
...
+                sources.append(newSource)
+            }
+        }
+        ObserverUtils.forEach(observers: videoObservers) { (observer: AudioVideoObserver) in
+            observer.remoteVideoSourcesDidBecomeAvailable(sources: sources)
+        }
+    }
+    
+    public func remoteVideoSourcesDidBecomeUnavailable(_ sourcesInternal: [RemoteVideoSourceInternal]) {
+        var sources = [RemoteVideoSource]()
```
Comment (hokyungh) [inline AmazonChimeSDK/AmazonChimeSDK/internal/video/DefaultVideoClientController.swift:248]: same as above

**Review by hokyungh — COMMENTED** [review]
Were you able to test it on Objective C just to make sure if it is compatible?

**Comment by ziyiz-amzn** [comment]
> Were you able to test it on Objective C just to make sure if it is compatible?

I have not tried. Will do.

File: AmazonChimeSDK/AmazonChimeSDK/internal/video/DefaultVideoClientController.swift [hunk 4]
```diff
@@ -368,6 +396,30 @@ extension DefaultVideoClientController: VideoClientController {
...
+        }
+        logger.info(msg: "Updating video subscriptions")
+        
+        let addedOrUpdatedInternal = Dictionary(uniqueKeysWithValues:
+            addedOrUpdated.map { source, config in
+                (RemoteVideoSourceInternal(attendeeId: source.attendeeId),
+                 VideoSubscriptionConfigurationInternal(
+                    priority: PriorityInternal(rawValue: UInt(config.priority.rawValue)) ?? PriorityInternal.highest,
+                    targetResolution: ResolutionInternal.init(width: Int32(config.resolution.width),
+                                                              height: Int32(config.resolution.height))))
```
Comment (dylonChime) [inline AmazonChimeSDK/AmazonChimeSDK/internal/video/DefaultVideoClientController.swift:413]: Just curious, what happens if  `config.resolution.width` and `config.resolution.height` are both zero or negative? Do we need to check for valid resolutions?

File: AmazonChimeSDKDemo/AmazonChimeSDKDemo/MeetingModel.swift [hunk 5]
```diff
@@ -431,6 +438,24 @@ extension MeetingModel: AudioVideoObserver {
     func videoSessionDidStartConnecting() {
         logWithFunctionName()
     }
+    
+    func remoteVideoSourcesDidBecomeAvailable(sources: [RemoteVideoSource]) {
+        logWithFunctionName()
+        var added = [RemoteVideoSource: VideoSubscriptionConfiguration]()
+        sources.forEach { source in
+            added[source] = VideoSubscriptionConfiguration()
+            videoModel.remoteVideoSourceConfigurations[source.attendeeId] = VideoSubscriptionConfiguration()
```
Comment (dylonChime) [inline AmazonChimeSDKDemo/AmazonChimeSDKDemo/MeetingModel.swift:447]: let videoSubscriptionConfiguration = VideoSubscriptionConfiguration()
added[source] = videoSubscriptionConfiguration
videoModel.remoteVideoSourceConfigurations[source.attendeeId] = videoSubscriptionConfiguration

**Comment by hensmi-amazon** [comment]
@dylonChime these are just converted to a single number for target bitrates internally. Poor choice of values would only at worst cause lower receive resolutions the desired, so I'm not to concerned here.

**Review by hokyungh — APPROVED** [review]

**Review by dylonChime — APPROVED** [review]

**Pull request merged**